integration/test/Testlib/PTest.hs

```diff
@@ -15,38 +19,96 @@ class HasTests x where
 instance HasTests (App ()) where
   mkTests m n s f x = [(m, n, s, f, x)]
 
-instance HasTests x => HasTests (Domain -> x) where
+instance (HasTests x, TestCases a) => HasTests (a -> x) where
   mkTests m n s f x =
-    mkTests m (n <> "[domain=own]") s f (x OwnDomain)
-      <> mkTests m (n <> "[domain=other]") s f (x OtherDomain)
+    flip foldMap (testCases @a) \tc ->
+      mkTests m (n <> tc.testCaseName) s f (x tc.testCase)
```

Comment:
i'd prefer for this kind of refactoring to happen in separate PRs next time, but this is really nice!